README.md
Overview
--------

This is a deployment and console wrapper for Minecraft which allows for more
convenient administration of a Minecraft server.

Installation
------------

    gem install minecraft

This will install the Minecraft extension library and binary, the Minecraft
server jarfile will be downloaded from http://minecraft.net when the binary is
ran for the first time.

Usage
-----

    mkdir ~/MinecraftServer
    cd ~/MinecraftServer
    minecraft

    minecraft -h

Current Features
----------------

Please refer to `Minecraft::Extensions#initialize`.

    !give <item> <quantity>       # Gives the user an <item>, the <quantity> 
                                  # upper bound is 2560 unlike the Minecraft 
                                  # default of 64.  Items can be specified by
                                  # approximate name.
    !giveall <item> <quantity>    # Gives all connected users an <item>.
    !tp <target_user>             # Teleports the user to the <target_user>.
    !tpall                        # Teleports all connected user to the user.
    !nom                          # Give a golden apple the user.
    !nomall                       # Give a golden apple to every connected user.
    !kit <group>                  # Gives a kit to the user.
    !kitall <group>               # Give a kit to every connected user.
    !kitlist                      # Lists the available kits.
    !addtimer <item> <frequency>  # Adds a timer to give <item> to the user ever <frequency> seconds.
    !deltimer <item>              # Delete a timer associated with <item>.
    !printtimer                   # Print timers of the user.
    !printtime                    # Print the current counter time.
    !list                         # List the connected users, notes ops.
    !property <key>               # Check a server property by key.
    !property                     # List server properties available.
    !s <label> <command>          # Associates a command to a shortcut label for the user.
    !s <label>                    # Runs the command with the associated label.
    !shortcuts                    # Lists the users shortcuts.
    !rules                        # Prints the server rules.
    !uptime <user>                # Prints the uptime of the given user.
    !uptime                       # Prints the uptime of the current user.
    !help                         # Outputs the help contents.
    !hop <user>                   # Give a user half-op privileges.
    !dehop <user>                 # Remove a users half-op privileges.
    !morning                      # Change time of day to morning.
    !evening                      # Change time of day to evening.
    !day                          # Change time of day to daytime.
    !night                        # Change time of day to night.
    !dawn                         # Change time of day to dawn.
    !dusk                         # Change time of day to dusk.
    !roulette                     # Kick a random person, person requesting has a higher chance.
    !kickvote <user>              # Initiate or vote for a kickvote against a user.
    !kickvote                     # Vote to kick the last initiated user.
    !vote                         # Vote to kick the last initiated user.
    !cancelvote <user>            # Cancel to kickvote on a user.
    !points <user> <quantity>     # Give a user points.
    !board <user>                 # Check a users points.
    !board                        # View the leaderboard of points.
    !om <noms>                    # Give golden apples equivalent to the number of noms.
    !warptime                     # Prints the current time rate.
    !warptime <rate>              # Adds <rate> seconds every ten seconds to time.
    !stop                         # Stops all the users timers.
    !welcome                      # Changes the welcome message during runtime.
    !memo <user> <message>        # Leaves a memo for the user.
    !disco                        # Turns time into a dancefloor.
    !dnd                          # Toggles the users do-not-disturb status.
    !disturb <user>               # An op can remove a user from the DND list.
    !printdnd                     # Prints the list of do-not-disturbed users.
    !todo                         # Prints the list of items todo.
    !todo <item>                  # Adds a todo list item.
    !finished <item>              # Removes an item from the todo list.

Development Path
----------------

### Major Release Milestone

- Complete test coverage
- Complete documentation coverage
- All feature requests closed
- All bugs closed
- Thorough testing done across multiple environments


### Semantic Versioning

- http://semver.org
- Currently minor and patch level releases are being made, nothing is stable
yet.
- Once the repository is completely stable and I am satisified with the feature
set, a major version will be released.  Any changes afterwards are backwards
compatible until the major version is incremented.

Contributors
------------

Forks are welcomed, pull requests will be prompty reviewed!

- Ian Horsman

Pull Request Guide
------------------

Pull requests are currently being accepted, please make sure that you fork off
of the `development` branch, preferably then developing on a feature branch.

Make sure your code includes tests and follow similar methodologies used in the
the code base, make sure to read relevant chunks of the code base before
committing.  Pull requests will be merged very strictly.

Read [https://github.com/basicxman/minecraft/wiki/Writing-tests](here) for how
we develop code with test driven development.

For any non-straightforward functionality (especially command naming and
syntax) you are encouraged to open up an
[http://github.com/basicxman/minecraft/issues](issue) for discussion, issues
will be read and commented on promptly.  Remember to report any bugs as well,
a bug can be something as simple as a command not returning feedback to the
user.


Notice
------

Minecraft is copyright of Mojang AB and developed by Markus Persson (@notch).


CHANGES.md
0.3.3 (2011-09-03)
------------------

* Refactored tests.
* Word wrapped more commands (should be all long outputs now).
* Source file reading bug fixed.
* Fixed an `all` style command bug.

0.3.2 (2011-08-31)
------------------

* Added dye kits with give support..
* Added a tool kit.
* Implemented command history.

0.3.1 (2011-08-30)
------------------

* Misc bug fixes.
* Revised README.
* Removed two time quotes.

0.3.0 (2011-08-29)
------------------

* Added `!warptime`, `!stop`, `!welcome`, `!memo`, `!disco`, `!dnd`, `!disturb`, `!printdnd`, `!todo`, `!finished`.
* Moved time commands out of `method_missing`.
* Added capability for users to not be disturbed by teleporting or `all` commands.
* Removed `add_command` calls, now indexes information from source code on initialization.
* Added more documentation (back at 100% coverage).
* Added many more test suites.
* Misc. bug fixes including lots of command feedback.
* Shortcuts and kits can now be accessed with `:` instead of `!s` or `!kit`.
* `!help` is now dynamic and includes command-specific help.
* Coloured terminal output.
* Basic word wrapped output.

0.2.1 (2011-08-22)
------------------

* Fixed `!points` and `!board`
* Bugfix on `call_command`
* Added more test suites.
* A `!help` command suitable for new privilege system.

0.2.0 (2011-08-21)
------------------

* Added configuration file capability.
* Added `!points`, `!board`, `!kickvote`, `!vote`, `!cancelvote`, `!kickvotes`, `!roulette`, `change_time`, `!om`, `!dehop`, `!hop`
* Added time changes (morning, night, dawn, etc...)
* Separated half-op and op privileges.
* `!property` now lists properties if none specified.
* Bugfix for `all` style commands (makes sure `:all => true`) is specified.
* `switch` is an alias for `lever`
* Refactored file saving/loading.
* Bugfix for default CLI options.
* Now saves instance files as well as executes `save-all` periodically.
* Bugfix for `uptime` stuff, offline users now can be computed.

0.1.0 (2011-08-20)
------------------

* Added yadocs.
* Updated README.
* Added test suites.

0.0.5 (2011-08-17)
------------------

* Added `savefreq` and `welcome_message` CLI options.
* Refactored `Minecraft` module, added a `Runtime` class.
* Added `!printtime`, `!printtimer`, `!s`, `!shortcuts` commands.
* Clarified console information/error messages.
* Fixed kits.
* Saves timers and shortcuts.
* Displays welcome message to connecting users.
* Major refactoring to server process handling.
* `save_all` on exit works now, proper exit handling.  No longer used `Thread#join`
* Toggling mob state (`--tempmob`) prints new state.

0.0.4 (2011-08-14)
------------------

* Added the `!rules` command and associated `rules` CLI option.
* Refactored `give` logic.
* `!property`, `!uptime` commands added.
* More intelligent quantifiers.
* Monitors kicks and bans.

0.0.3 (2011-08-13)
------------------

* Split command methods into the `Minecraft::Commands` module.
* Validates `!kit` command.
* Added `!list`, `!addtimer`, `!deltimer`, `!printtimer`, `!kitlist` commands.
* Approximate item names implemented.
* Fixed item naming bugs.
* Major refactoring done to commands.
* Logs user uptime.
* Catches processing exceptions.

0.0.2 (2011-08-13)
------------------

* Command line options added.
* Trapped the interrupt signal for an exit hook.
* `no_run`, `update`, `min_memory`, `max_memory`, `no_auto_save`, `tempmobs` CLI options implemented.
* Fixed indentation.
* `!giveall`, `!tpall`, `!help` commands implemented.
* Meta checks in place (ops, join/part).
* Privilege errors.
* Refactored `Minecraft::Server`, added `Minecraft::Tools`.

0.0.1 (2011-08-12)
------------------

* `:diamond`, `:garmour`, `:armour`, `:ranged`, `:nether`, `:portal` kits.
* Data values hash in place.
* Simple console processing in place.
* `!give`, `!kit`, `!tp`, `!nom` commands implemented.


test/helper.rb
unless Object.const_defined? 'Minecraft'
  $:.unshift File.expand_path("../../lib", __FILE__)
  require "minecraft"
end

require "minitest/autorun"
require "fileutils"
require "stringio"
require "turn"

module Minecraft
  class Extensions
    attr_accessor :commands, :users, :ops, :hops, :counter, :server, :userkickvotes, :last_kick_vote, :useruptime, :timers, :usershortcuts, :userlog, :userpoints, :vote_threshold, :userdnd, :welcome_message, :memos, :todo_items, :command_history
  end
end

class Test < MiniTest::Unit::TestCase
  # Standard test.
  def self.test(name, &block)
    define_method("test_#{name.gsub(/\W/, '_')}", block)
  end

  # Sandbox test which creates a sandboxed environment for a mocked Minecraft
  # server to run in.
  def self.sandbox_test(name, &block)
    p = Proc.new do
      FileUtils.mkdir("mc") unless File.exists? "mc"
      FileUtils.cd("mc") do
        FileUtils.touch("ops.txt")
        FileUtils.touch("server.properties")
        FileUtils.rm_f("command_history.json")
        instance_eval(&block)
      end
    end
    define_method("test_#{name.gsub(/\W/, '_')}", p)
  end

  def ext(opts = {})
    @ext = Minecraft::Extensions.new(StringIO.new, {})
    @ext.ops  = opts[:ops]  || []
    @ext.hops = opts[:hops] || []
    @ext.users = @ext.ops + @ext.hops + (opts[:users] || [])
    @ext.users.uniq!
  end

  def call(string)
    @ext.call_command(*string.split(" "))
  end

  def output
    @ext.server.string
  end

  def output_lines
    output.split("\n")
  end

  def output_line
    output.gsub("\n", "")
  end

  def clear
    @ext.server.string = ""
  end
end


test/tools_test.rb
require "helper"
require "minecraft/tools"

class ToolsTest < Test
  test "should get minecraft download url" do
    url = Minecraft::Tools.get_minecraft_page
    refute_nil url.index "minecraft_server.jar"
  end

  test "should generate a proper default command" do
    command = Minecraft::Tools.command({})
    assert_equal "java -Xmx1024M -Xms1024M -jar minecraft_server.jar nogui", command
  end

  test "should properly modify server properties when toggling mobs" do
    File.open("server.properties", "w") do |f| f.print <<-eof
#Minecraft server properties
#Thu May 26 15:19:31 EDT 2011
view-distance=10
spawn-monsters=false
online-mode=true
eof
    end
    assert_equal "true", Minecraft::Tools.toggle_mobs
    refute_nil File.read("server.properties").index("spawn-monsters=true")
    assert_equal "false", Minecraft::Tools.toggle_mobs
    refute_nil File.read("server.properties").index("spawn-monsters=false")
    FileUtils.rm_f("server.properties")
  end
end


lib/minecraft/tools.rb
require "net/http"

module Minecraft
  # Methods for external manipulation of the current deployment.
  module Tools
    # Checks if minecraft_server.jar and calls the download method if not.
    def self.check_jarfile
      download_minecraft unless File.exists? "minecraft_server.jar"
    end

    # Downloads the minecraft server jarfile into the current directory.
    def self.download_minecraft
      url = get_minecraft_page
      puts "[+] Downloading Minecraft server..."
      `wget http://minecraft.net/#{url} -O minecraft_server.jar -q`
    end

    # Parses the miencraft.net download page for the current jarfile URL.
    def self.get_minecraft_page
      page = Net::HTTP.get("www.minecraft.net", "/download.jsp")
      data = page.match /\"([0-9a-zA-Z_\/]*minecraft_server\.jar\?v=[0-9]+)/
      data[1]
    end

    # Generates a command for running the server with default settings including memory.
    #
    # @param [Slop] opts Command line options from Slop, used for memory specification.
    def self.command(opts)
      "java -Xmx#{opts[:max_memory] || "1024M"} -Xms#{opts[:min_memory] || "1024M"} -jar minecraft_server.jar nogui"
    end

    # Toggles mobs in server.properties and returns the new state.
    def self.toggle_mobs
      return unless File.exists? "server.properties"
      content = File.read("server.properties")
      state = content.match(/spawn\-monsters=(true|false)/)[1]
      new_state = state == "true" ? "false" : "true"
      content.gsub! "spawn-monsters=#{state}", "spawn-monsters=#{new_state}"

      File.open("server.properties", "w") { |f| f.print content }
      return new_state
    end

    # Grabs the extension configuration file and parses it.
    #
    # @return [Hash] Configuration hash.
    def self.get_configuration_file
      return {} unless File.exists? "minecraft.properties"
      File.readlines("minecraft.properties").map { |l| l.split(" ") }.inject({}) do |hash, (key, *value)|
        hash.merge({ key.to_sym => config_value(value) })
      end
    end

    # Parses a value (or set of) from a configuration file.
    #
    # @param [String] An array of values.
    # @return [String] The string value joined together.
    # @return [Boolean] Will return a boolean true if the value is blank (meant
    # to be a boolean configuration flag).
    def self.config_value(value)
      return true if value.nil? or value.length == 0
      return value.join(" ")
    end
  end
end


lib/minecraft/server.rb
require "open3"

module Minecraft
  # An instance of the server class will start a Java subprocess running the
  # Minecraft server jarfile.  The interrupt signal will be trapped to run an
  # exit hook and threads will be created to process all pipes.
  class Server
    attr_accessor :sin

    # New Server instance.
    #
    # @param [String] command The command for the subprocess.
    # @param [Slop] opts Command line options from Slop.
    # @example
    #   opts = Slop.new do
    #     ...
    #   end
    #   server = Server.new("java -jar minecraft_server.jar", opts)
    def initialize(command, opts)
      @sin, @sout, @serr, @thr = Open3.popen3(command)
      @extensions = Extensions.new(@sin, opts)
      @opts = opts

      trap("SIGINT") { minecraft_exit }

      @threads = []
      @threads << Thread.new { loop { @extensions.process(@sout.gets) } }
      @threads << Thread.new { loop { @extensions.process(@serr.gets) } }
      @threads << Thread.new { loop { @extensions.periodic; sleep 1 } }
      @threads << Thread.new { loop { @sin.puts $stdin.gets } }
      @thr.value
      exit!
    end
    
    # An exit hook, checks if mobs need to be untoggled and saves the server.
    # Server is stopped gracefully.
    def minecraft_exit
      puts "[+] Restoring previous mob state to #{Minecraft::Tools.toggle_mobs}." if @opts[:tempmobs]
      puts "[~] The current welcome message is:"
      puts @extensions.welcome_message
      puts "\n[+] Saving..."
      @extensions.save
      @threads.each(&:kill)
      @sin.puts("save-all")
      @sin.puts("stop")
    end
  end
end


lib/minecraft/runtime.rb
module Minecraft
  # An instance of the Runtime class will spool up the server and do any
  # preliminary work.
  class Runtime
    # New Runtime instance.
    #
    # @param [Slop] opts Command line options from slop.
    # @example
    #   opts = Slop.parse do
    #     ...
    #   end
    #   runtime = Runtime.new(opts)
    def initialize(opts)
      @opts = opts
      pre_checks

      Minecraft::Tools.check_jarfile
      unless @opts[:no_run]
        command = Minecraft::Tools.command(@opts)
        puts "[+] #{command}"
        server = Minecraft::Server.new(command, @opts)
        server.sin.puts("save-on") unless @opts[:no_auto_save]
      end
    end

    # Checks if Minecraft needs to be updated, checks if mobs are to be
    # toggled.
    def pre_checks
      Minecraft::Tools.download_minecraft if @opts[:update]
      puts "[+] Temporarily toggling mobs.  Setting to #{Minecraft::Tools.toggle_mobs}." if @opts[:tempmobs]
    end
  end
end


lib/minecraft/data.rb
module Minecraft
  # Contains kit and item hashes takes care of buckets for items.
  module Data
    # Item kits.
    KITS = {
      :diamond    => [276, 277, 278, 279, 293],
      :goldarmour => [314, 315, 316, 317],
      :armour     => [310, 311, 312, 313],
      :tools      => [277, 278],
      :ranged     => [261, [262, 320]],
      :nether     => [261, [262, 320], [89, 128], 278, 276],
      :portal     => [[49, 14], 259],
      :redstone   => [[331, 256], [356, 64], [69, 6], [77, 6], [70, 6], [76, 64], [29, 24], [4, 64]],
      :circuits   => [[331, 256], [356, 64], [76, 64], [4, 64]],
      :cannon     => [[356, 5], [331, 18], [76, 2], [4, 64], [46, 14], 77, [8, 2]],
      :blacksmith => [61, [263, 16], [264, 3], [265, 32], [266, 16]],
      :miner      => [278, [50, 64], 277, [65, 64]],
      :farmer     => [293, [295, 128], 277],
      :chef       => [[297, 3], [296, 192], [320, 3], 357, [353, 64], 354, [350, 2]],
      :lightgray  => [[351, 64], [352, 64]],
      :gray       => [[351, 64], [352, 64]],
      :black      => [[351, 64]],
      :red        => [[38, 64]],
      :orange     => [[38, 64], [37, 64]],
      :yellow     => [[37, 64]],
      :lime       => [[352, 64], [81, 64]],
      :green      => [[81, 64]],
      :lightblue  => [[352, 64], [22, 64]],
      :cyan       => [[22, 64], [81, 64]],
      :blue       => [[22, 64]],
      :purple     => [[22, 64], [38, 64]],
      :magenta    => [[22, 64], [38, 64], [38, 64], [352, 64]],
      :pink       => [[352, 64], [38, 64]]
    }

    # Possible wool colours.
    WOOL_COLOURS = %w(
      lightgray
      gray
      black
      red
      orange
      yellow
      lime
      green
      lightblue
      cyan
      blue
      purple
      magenta
      pink
    )

    # Values for time, 0 to 24000.  0 is dawn, 12000 is dusk.
    TIME = {
      :morning => 1000,
      :evening => 13000,
      :day     => 5000,
      :night   => 16000,
      :dawn    => 0,
      :dusk    => 12000
    }

    # Quotes for each time of day.
    TIME_QUOTES = {
      :morning => "",
      :evening => "Let them come.",
      :day     => "",
      :night   => "",
      :dawn    => "",
      :dusk    => ""
    }

    # All item keys.
    DATA_VALUE_HASH = {
      "glass"                   => "20",
      "purple wool"             => "35",
      "obsidian"                => "49",
      "redstone ore"            => "73",
      "stone"                   => "1",
      "lapis lazuli ore"        => "21",
      "blue wool"               => "35",
      "torch"                   => "50",
      "glowing redstone ore"    => "74",
      "grass"                   => "2",
      "lapis lazuli block"      => "22",
      "brown wool"              => "35",
      "fire"                    => "51",
      "dirt"                    => "3",
      "dispenser"               => "23",
      "dark green wool"         => "35",
      "monster spawner"         => "52",
      "cobblestone"             => "4",
      "sandstone"               => "24",
      "red wool"                => "35",
      "wooden stairs"           => "53",
      "stone button"            => "77",
      "wooden plank"            => "5",
      "note block"              => "25",
      "black wool"              => "35",
      "chest"                   => "54",
      "snow"                    => "78",
      "sapling"                 => "6",
      "bed"                     => "355",
      "yellow flower"           => "37",
      "redstone wire"           => "55",
      "ice"                     => "79",
      "spruce sapling"          => "6",
      "powered rail"            => "27",
      "red rose"                => "38",
      "diamond ore"             => "56",
      "snow block"              => "80",
      "birch sapling"           => "6",
      "detector rail"           => "28",
      "brown mushroom"          => "39",
      "diamond block"           => "57",
      "cactus"                  => "81",
      "bedrock"                 => "7",
      "sticky piston"           => "29",
      "red mushroom"            => "40",
      "crafting table"          => "58",
      "clay block"              => "82",
      "water"                   => "8",
      "cobweb"                  => "30",
      "gold block"              => "41",
      "crops "                  => "59",
      "sugar cane"              => "338",
      "stationary water"        => "9",
      "tall grass"              => "31",
      "iron block"              => "42",
      "farmland "               => "60",
      "jukebox"                 => "84",
      "lava"                    => "10",
      "dead shrubs"             => "32",
      "double stone slab"       => "43",
      "furnace"                 => "61",
      "fence"                   => "85",
      "stationary lava "        => "11",
      "piston"                  => "33",
      "double sandstone slab"   => "43",
      "burning furnace"         => "62",
      "pumpkin"                 => "86",
      "sand"                    => "12",
      "white wool"              => "35",
      "double wooden slab"      => "43",
      "sign post"               => "63",
      "netherrack"              => "87",
      "gravel"                  => "13",
      "orange wool"             => "35",
      "double cobblestone slab" => "43",
      "wooden door"             => "324",
      "soul sand"               => "88",
      "gold ore"                => "14",
      "magenta wool"            => "35",
      "stone slab"              => "44",
      "ladder"                  => "65",
      "glowstone"               => "89",
      "glowstone block"         => "89",
      "iron ore"                => "15",
      "light blue wool"         => "35",
      "sandstone slab"          => "44",
      "rails"                   => "66",
      "portal"                  => "90",
      "coal ore"                => "16",
      "yellow wool"             => "35",
      "wooden slab"             => "44",
      "cobblestone stairs"      => "67",
      "jack-o-lantern"          => "91",
      "jack o lantern"          => "91",
      "oak wood"                => "17",
      "light green wool"        => "35",
      "cobblestone slab"        => "44",
      "wall sign"               => "68",
      "cake block"              => "92",
      "pine wood"               => "17",
      "pink wool"               => "35",
      "brick block"             => "45",
      "lever"                   => "69",
      "switch"                  => "69",
      "birch wood"              => "17",
      "gray wool"               => "35",
      "tnt"                     => "46",
      "stone pressure plate"    => "70",
      "leaves"                  => "18",
      "light gray wool"         => "35",
      "bookshelf"               => "47",
      "iron door"               => "330",
      "locked chest"            => "95",
      "sponge"                  => "19",
      "cyan wool"               => "35",
      "moss stone"              => "48",
      "wooden pressure plate"   => "72",
      "trapdoor"                => "96",
      "iron shovel"             => "256",
      "stick"                   => "280",
      "chainmail boots"         => "305",
      "lapis lazuli"            => "351",
      "iron pickaxe"            => "257",
      "bowl"                    => "281",
      "iron helmet"             => "306",
      "redstone"                => "331",
      "purple dye"              => "351",
      "iron axe"                => "258",
      "mushroom soup"           => "282",
      "iron chestplate"         => "307",
      "snowball"                => "332",
      "cyan dye"                => "351",
      "flint and steel"         => "259",
      "gold sword"              => "283",
      "iron leggings"           => "308",
      "boat"                    => "333",
      "light gray dye"          => "351",
      "apple"                   => "260",
      "gold shovel"             => "284",
      "iron boots"              => "309",
      "leather"                 => "334",
      "gray dye"                => "351",
      "bow"                     => "261",
      "gold pickaxe"            => "285",
      "diamond helmet"          => "310",
      "milk"                    => "335",
      "pink dye"                => "351",
      "arrow"                   => "262",
      "gold axe"                => "286",
      "diamond chestplate"      => "311",
      "clay brick"              => "336",
      "lime dye"                => "351",
      "coal"                    => "263",
      "string"                  => "287",
      "diamond leggings"        => "312",
      "clay balls"              => "337",
      "dandelion yellow"        => "351",
      "charcoal"                => "263",
      "feather"                 => "288",
      "diamond boots"           => "313",
      "light blue dye"          => "351",
      "diamond"                 => "264",
      "gunpowder"               => "289",
      "gold helmet"             => "314",
      "paper"                   => "339",
      "magenta dye"             => "351",
      "iron ingot"              => "265",
      "wooden hoe"              => "290",
      "gold chestplate"         => "315",
      "book"                    => "340",
      "orange dye"              => "351",
      "gold ingot"              => "266",
      "stone hoe"               => "291",
      "gold leggings"           => "316",
      "slimeball"               => "341",
      "bone meal"               => "351",
      "iron sword"              => "267",
      "iron hoe"                => "292",
      "gold boots"              => "317",
      "storage minecart"        => "342",
      "bone"                    => "352",
      "wooden sword"            => "268",
      "diamond hoe"             => "293",
      "flint"                   => "318",
      "powered minecart"        => "343",
      "sugar"                   => "353",
      "wooden shovel"           => "269",
      "gold hoe"                => "294",
      "raw porkchop"            => "319",
      "egg"                     => "344",
      "cake"                    => "354",
      "wooden pickaxe"          => "270",
      "seeds"                   => "295",
      "cooked porkchop"         => "320",
      "compass"                 => "345",
      "wooden axe"              => "271",
      "wheat"                   => "296",
      "paintings"               => "321",
      "fishing rod"             => "346",
      "redstone repeater"       => "356",
      "stone sword"             => "272",
      "bread"                   => "297",
      "golden apple"            => "322",
      "clock"                   => "347",
      "cookie"                  => "357",
      "stone shovel"            => "273",
      "leather helmet"          => "298",
      "sign"                    => "323",
      "glowstone dust"          => "348",
      "map"                     => "358",
      "stone pickaxe"           => "274",
      "leather chestplate"      => "299",
      "raw fish"                => "349",
      "shears"                  => "359",
      "stone axe"               => "275",
      "leather leggings"        => "300",
      "bucket"                  => "325",
      "cooked fish"             => "350",
      "gold music disc"         => "2256",
      "diamond sword"           => "276",
      "leather boots"           => "301",
      "water bucket"            => "326",
      "ink sac"                 => "351",
      "green music disc"        => "2257",
      "diamond shovel"          => "277",
      "chainmail helmet"        => "302",
      "lava bucket"             => "327",
      "rose red"                => "351",
      "diamond pickaxe"         => "278",
      "chainmail chestplate"    => "303",
      "minecart"                => "328",
      "cactus green"            => "351",
      "diamond axe"             => "279",
      "chainmail leggings"      => "304",
      "saddle"                  => "329",
      "cocoa beans"             => "351"
    }

    # Alphabetical buckets for quicker approximate matching.  Filled on init.
    ITEM_BUCKETS = {}
    DATA_VALUE_HASH.each_key do |key|
      bucket = key[0]
      ITEM_BUCKETS[bucket] ||= []
      ITEM_BUCKETS[bucket] << key
    end
  end
end
